File: kevoree-library/javase/org.kevoree.library.javase.nanohttp/src/main/scala/org/kevoree/library/javase/nanohttp/NodeNetworkHelper.scala
package org.kevoree.library.javase.nanohttp

import org.kevoree.ContainerRoot
import scala.collection.JavaConversions._
import java.net.{SocketException, NetworkInterface}
import org.slf4j.LoggerFactory
import org.kevoree.api.service.core.script.KevScriptEngineFactory
import org.kevoree.framework.{KevoreePlatformHelper, AbstractGroupType}

/**
 * User: Erwan Daubert - erwan.daubert@gmail.com
 * Date: 01/03/12
 * Time: 14:46
 *
 * @author Erwan Daubert
 * @version 1.0
 */

object NodeNetworkHelper {
  private val logger = LoggerFactory.getLogger(getClass)


  def updateModelWithNetworkProperty (group: AbstractGroupType) : Option[ContainerRoot] = {
    val ipObject = group.getDictionary.get("ip")
    if (ipObject != null && ipObject.toString != "" && ipObject.toString != "0.0.0.0") {
      addNetworkProperty(group.getModelService.getLastModel, group.getNodeName, Array[(String, String)]((ipObject.toString, "unknown")), group.getKevScriptEngineFactory)
    } else {
      val addresses = getAddresses
      if (addresses.length > 0) {
        /*val modelOption = */addNetworkProperty(group.getModelService.getLastModel, group.getNodeName, addresses, group.getKevScriptEngineFactory)
        /*if (modelOption.isDefined) {
          group.getModelService.updateModel(modelOption.get)
        }*/
      } else {
        None
      }
    }
  }

  /*def addNetworkProperty (model: ContainerRoot, nodeName: String, ips: Array[(String, String)], kevScriptEngineFactory :  KevScriptEngineFactory): Option[ContainerRoot] = {
    val newModel = (new ModelCloner).clone(model)

    val scriptBuilder = new StringBuilder
    ips.foreach {
      ip =>
        scriptBuilder append
          "network " + nodeName + " {\"" + Constants.KEVOREE_PLATFORM_REMOTE_NODE_IP + "\" = \"" + ip._1 + "\", \"" + Constants.KEVOREE_PLATFORM_REMOTE_NETWORK_TYPE + "\"=\"" + ip._2 + "\" }\n"
    }

    try {
      val kevScripEngine = kevScriptEngineFactory.createKevScriptEngine (newModel)
      kevScripEngine.append(scriptBuilder.toString())
      Some(kevScripEngine.interpret())
    } catch {
      case _@e => None
    }
  }*/

  def getAddresses: Array[(String, String)] = {
    var addresses = Array[(String, String)]()
    try {
      NetworkInterface.getNetworkInterfaces.foreach {
        networkInterface =>
          if (!networkInterface.isLoopback) {
            networkInterface.getInterfaceAddresses.foreach {
              interfaceAddress =>
                addresses = addresses ++ Array[(String, String)]((interfaceAddress.getAddress.getHostAddress, networkInterface.getDisplayName))
            }
          }
      }
    }
    catch {
      case ex: SocketException => {
        logger.info("Unable to retrieve IP addresses", ex)
      }
    }
    addresses
  }

  def addNetworkProperty (model: ContainerRoot, nodeName: String, ips: Array[(String, String)], kevScriptEngineFactory: KevScriptEngineFactory): Option[ContainerRoot] = {
    ips.foreach {
      ip =>
        KevoreePlatformHelper.updateNodeLinkProp(model, nodeName, nodeName, org.kevoree.framework.Constants.KEVOREE_PLATFORM_REMOTE_NODE_IP, ip._1, ip._2, 100)
        logger.info("add {} as IP of {}", ip._1, nodeName)
    }
    Some(model)
  }

}

File: kevoree-corelibrary/javase/org.kevoree.library.javase.jexxus/src/main/java/org/kevoree/library/NodeNetworkHelper.java
package org.kevoree.library;

import org.kevoree.ContainerRoot;
import org.kevoree.DictionaryAttribute;
import org.kevoree.api.service.core.script.KevScriptEngineFactory;
import org.kevoree.cloner.ModelCloner;
import org.kevoree.framework.AbstractGroupType;
import org.kevoree.framework.KevoreePlatformHelper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.Inet4Address;
import java.net.InterfaceAddress;
import java.net.NetworkInterface;
import java.util.Enumeration;

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 09/11/12
 * Time: 07:41
 */
public class NodeNetworkHelper {

    private static Logger logger = LoggerFactory.getLogger(NodeNetworkHelper.class);

    public static void main(String[] args){
        getAddresses();

    }

    public static java.util.HashMap<String,String> getAddresses() {
        java.util.HashMap<String,String> addresses = new java.util.HashMap<String,String>();
        try {
            Enumeration<NetworkInterface> networkInterfaces = NetworkInterface.getNetworkInterfaces();
            while(networkInterfaces.hasMoreElements()){
                NetworkInterface networkInterface = networkInterfaces.nextElement();
                if (!networkInterface.isLoopback()) {
                      for(InterfaceAddress interfaceAddress : networkInterface.getInterfaceAddresses()){
                          if(interfaceAddress.getAddress() instanceof Inet4Address){
                              addresses.put(interfaceAddress.getAddress().getHostAddress(),networkInterface.getDisplayName());
                          }
                      }
                }
            }
        }
        catch(Exception e) {
            logger.error("",e);
        }
        return addresses;
    }


    public static ContainerRoot addNetworkProperty (ContainerRoot model, String nodeName ,java.util.HashMap<String,String> ips , KevScriptEngineFactory kevScriptEngineFactory) {
        for(String key : ips.keySet()){
            KevoreePlatformHelper.updateNodeLinkProp(model, nodeName, nodeName, org.kevoree.framework.Constants.KEVOREE_PLATFORM_REMOTE_NODE_IP(), key, ips.get(key), 100);
            logger.info("add {} as IP of {}", key, nodeName);

        }
        return model;
    }

    private static ModelCloner cloner = new ModelCloner();

    public static ContainerRoot updateModelWithNetworkProperty (AbstractGroupType group) {
        Object ipObject = group.getDictionary().get("ip");
        ContainerRoot readWriteModel = cloner.clone(group.getModelService().getLastModel());
        if (ipObject != null && !ipObject.toString().equals("") && !ipObject.toString().equals("0.0.0.0")) {
            java.util.HashMap<String,String> addresses = new java.util.HashMap<String,String>();
            addresses.put(ipObject.toString(),"unknown");
            return addNetworkProperty(readWriteModel, group.getNodeName(),addresses, group.getKevScriptEngineFactory());
        } else {
            java.util.HashMap<String,String> addresses = getAddresses();
            if (!addresses.isEmpty()) {
                return addNetworkProperty(readWriteModel, group.getNodeName(), addresses, group.getKevScriptEngineFactory());
            } else {
                return null;
            }
        }
    }

}
